parallel-install/pkg/preinstaller/resourceparser.go

```
@@ -0,0 +1,93 @@
+type GenericResourceParser struct {
+	decoder         runtime.Decoder
+}
+
+// NewGenericResourceParser returns a new instance of GenericResourceParser.
+func NewGenericResourceParser() *GenericResourceParser {
+	return &GenericResourceParser{
+		decoder:         initializeDecoder(),
+	}
+}
+
+func (c *GenericResourceParser) ParseUnstructuredResourceFrom(path string) (obj *unstructured.Unstructured, err error) {
+	manifest, err := ioutil.ReadFile(path)
+	if err != nil {
+		return obj, err
+	}
+
+	return c.parseResourceFrom(string(manifest))
+}
+
+func (c *GenericResourceParser) parseResourceFrom(manifest string) (*unstructured.Unstructured, error) {
+	var _, _, err = c.decoder.Decode([]byte(manifest), nil, nil)
+	if err != nil {
+		return nil, errors.New(fmt.Sprintf("Could not decode the resource file due to the following error: %s.", err.Error()))
+	}
+
+	converted, err := convertYamlToJson(manifest)
+	if err != nil {
+		return nil, errors.New(fmt.Sprintf("Could not convert the resource file to JSON due to the following error: %s.", err.Error()))
+	}
+
+	resource, err := parseManifest([]byte(converted))
+	if err != nil {
+		return nil, errors.New(fmt.Sprintf("Could not parse the resource file due to the following error: %s.", err.Error()))
+	}
+
+	return resource, nil
+}
+
+func convertYamlToJson(input string) (string, error) {
+	convertedInput, err := yaml.YAMLToJSON([]byte(input))
+	if err != nil {
+		return "", err
+	}
+	if string(convertedInput) != "null" {
+		return string(convertedInput), nil
+	}
+
+	return "", err
+}
+
+func parseManifest(input []byte) (*unstructured.Unstructured, error) {
+	var middleware map[string]interface{}
+	err := json.Unmarshal(input, &middleware)
+	if err != nil {
+		return nil, err
+	}
+
+	resource := &unstructured.Unstructured{
+		Object: middleware,
+	}
+	return resource, nil
+}
+
+func initializeDecoder() runtime.Decoder {
+	sch := runtime.NewScheme()
+	_ = clientgoscheme.AddToScheme(sch)
+	_ = apiextv1.AddToScheme(sch)
```

Please do not ignore errors.